src/semver/version.py

```diff
@@ -144,27 +162,122 @@ class Version:
+            if args and len(args) > 5:
+                raise ValueError("You cannot pass more than 5 arguments to Version")
+            elif len(args) > 1 and "." in str(args[0]):
+                raise ValueError(
+                    "You cannot pass a string and additional positional arguments"
+                )
+            allowed_types_in_args = (
+                (int, str, bytes),  # major
+                (int, str, bytes),  # minor
+                (int, str, bytes),  # patch
+                (str, bytes, int, type(None)),  # prerelease
+                (str, bytes, int, type(None)),  # build
+            )
+            return [
+                isinstance(item, allowed_types_in_args[i])
+                for i, item in enumerate(args)
+            ]
+
+        cls = self.__class__
+        verlist: List[Optional[StringOrInt]] = [None, None, None, None, None]
+
+        types_in_args = _check_types(*args)
+        if not all(types_in_args):
+            pos = types_in_args.index(False)
+            raise TypeError(
+                "not expecting type in argument position "
+                f"{pos} (type: {type(args[pos])})"
+            )
+        elif args and "." in str(args[0]):
+            # we have a version string as first argument
+            v = cls._parse(args[0])  # type: ignore
+            for idx, key in enumerate(
+                ("major", "minor", "patch", "prerelease", "build")
+            ):
+                verlist[idx] = v[key]
+        else:
+            for index, item in enumerate(args):
+                verlist[index] = args[index]  # type: ignore
+
         # Build a dictionary of the arguments except prerelease and build
-        version_parts = {"major": int(major), "minor": int(minor), "patch": int(patch)}
+        try:
+            version_parts = {
+                # Prefer major, minor, and patch arguments over args
+                "major": int(major or verlist[0] or 0),
+                "minor": int(minor or verlist[1] or 0),
+                "patch": int(patch or verlist[2] or 0),
+            }
+        except ValueError:
+            raise ValueError(
+                "Expected integer or integer string for major, minor, or patch"
+            )
 
         for name, value in version_parts.items():
             if value < 0:
                 raise ValueError(
                     "{!r} is negative. A version can only be positive.".format(name)
                 )
 
+        prerelease = cls._ensure_str(prerelease or verlist[3])  # type: ignore
+        build = cls._ensure_str(build or verlist[4])  # type: ignore
+
         self._major = version_parts["major"]
         self._minor = version_parts["minor"]
         self._patch = version_parts["patch"]
         self._prerelease = None if prerelease is None else str(prerelease)
         self._build = None if build is None else str(build)
 
+    @classmethod
+    def _ensure_str(cls, s: Optional[String], encoding="UTF-8") -> Optional[str]:
```

Comment:
```suggestion
    def _ensure_str(cls, s: Optional[StringOrInt], encoding="UTF-8") -> Optional[str]:
```
This should resolve type issues you might be having earlier when using that method.

Reply:
Thanks. I had to change the return type as well (to `Optional[StringOrInt]`), otherwise I get the error message:

    src/semver/version.py:228: error: Incompatible return value type (got "Union[str, int, None]", expected "Optional[str]")

Reply:
Make sure you use `Optional[str]` here, since it is meant to ensure that it is a `str`.

Reply:
> I think we may have an architecture problem on our hands here. We shouldn't be calling `_ensure_str` on things other than a variation of string, so we ought to make sure we aren't passing in an `int`.

Yes, could be. That's why I didn't merge this PR yet. It is a bigger change and I would like to be fully sure and discuss that with you first. I wouldn't like to introduce any regression. :wink: 

Well, I thought, it could be an elegant way to have this line:

```python
prerelease = cls._ensure_str(prerelease or verlist[3])
```

That would either convert bytes -> str or left None and int untouched. Further down this case is handled:

```python
self._prerelease = None if prerelease is None else str(prerelease)
```

So here is the final assignment to `self._prerelease` to either store None or a string (regardless if it was passed as bytes, str, or int).

Maybe the name of the function is a bit irritating. Or we need to find a better way to assign `self._prerelease` its final value. Any ideas?